### PR TITLE
Support doc comments for symbols declared off an anonymous type.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/12530

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide @dotnet/roslyn-compiler 

Roslyn-compiler, this PR adds a new API on Compilation to create anonymous type symbols.  This API is similar to hte recently added CreateTupleTypeSymbol and is needed by the IDE so that we can serialize out and recover anonymous type symbols for features like completion-list descriptions.

I have *not* written tests *yet*.  That's because i want feedback on the shape of the API and if i'm properly implementing it for VB and C# first.  Once there's consensus on how best to provide this API, i'm happy to go write suitable coverage.

--

Why is this needed?  Recently, our completion list code was refactored to move to a pure data model where we no longer hold onto rich objects (like symbols) anymore.  Completion list items that need symbolic information now created a serialized representation of teh symbol (using the SymbolKey type) and then resolve that key back to an actual symbol to support certain operations (like getting doc comments and quick info text).  

This works well for most symbol types, but fell over if the symbol contained an anonymous type in it.  That's because, during resolution, there was no way to get back to that anonymous type.  This regression in functionality was missed due to test coverage not being adequate enough when we made the completion change.

I've now added support for round-tripping this symbol type, and added tests to make sure it doesn't regress in the future.

File: src/Compilers/Core/Portable/Compilation/Compilation.cs [hunk 1]
```diff
@@ -851,6 +851,31 @@ public INamedTypeSymbol CreateTupleTypeSymbol(INamedTypeSymbol underlyingType, I
 
         protected abstract INamedTypeSymbol CommonCreateTupleTypeSymbol(INamedTypeSymbol underlyingType, ImmutableArray<string> elementNames);
 
+        /// <summary>
+        /// Returns a new anonymous type symbol with the given member types member names.
+        /// </summary>
+        public INamedTypeSymbol CreateAnonymousTypeSymbol(
+            ImmutableArray<ITypeSymbol> memberTypes, ImmutableArray<string> memberNames)
+        {
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Compilation/Compilation.cs]: Other potential signatures could be:

```CreateAnonymousTypeSymbol(ImmutableArray<ValueTuple<ITypeSymbol, string>> propertyDescriptions)``` or
```
struct AnonymousPropertyDescription
{
    public ITypeSymbol Type { get; }
    public string Name { get; }
}
...
CreateAnonymousTypeSymbol(ImmutableArray<AnonymousPropertyDescription> propertyDescriptions)
```

**Comment by CyrusNajmabadi** [comment]
test vsi please

**Comment by CyrusNajmabadi** [comment]
retest windows_vsi_p3_open_prtest please

**Comment by CyrusNajmabadi** [comment]
@dotnet-bot retest windows_vsi_p3_open_prtest please
// Previous failure: http://dotnet-ci.cloudapp.net/job/Private/job/dotnet_roslyn-internal/job/master/job/windows_vsi_p3_open_prtest/281/
// Retest reason:

**Comment by CyrusNajmabadi** [comment]
@dotnet-bot retest windows_vsi_p3_open_prtest please
// Previous failure: http://dotnet-ci.cloudapp.net/job/Private/job/dotnet_roslyn-internal/job/master/job/windows_vsi_p3_open_prtest/284/
// Retest reason:

**Comment by CyrusNajmabadi** [comment]
@dotnet-bot retest microbuild_prtest please
// Previous failure: http://dotnet-ci.cloudapp.net/job/dotnet_roslyn/job/master/job/microbuild_prtest/244/
// Retest reason:

**Comment by CyrusNajmabadi** [comment]
@dotnet-bot retest windows_vsi_p0_open_prtest please
// Previous failure: http://dotnet-ci.cloudapp.net/job/Private/job/dotnet_roslyn-internal/job/master/job/windows_vsi_p0_open_prtest/272/
// Retest reason:

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-compiler need your input/sign-off.  @dotnet/roslyn-ide need rview.

**Comment by CyrusNajmabadi** [comment]
@jcouv said i should tag @gafter directly. :)

**Comment by Pilchie** [comment]
:+1: for the IDE side.

File: src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs [hunk 2]
```diff
@@ -2835,6 +2835,20 @@ protected override INamedTypeSymbol CommonCreateTupleTypeSymbol(INamedTypeSymbol
...
 
+        protected override INamedTypeSymbol CommonCreateAnonymousTypeSymbol(
+            ImmutableArray<ITypeSymbol> memberTypes, ImmutableArray<string> memberNames)
+        {
+            for (int i = 0, n = memberTypes.Length; i < n; i++)
+            {
+                memberTypes[i].EnsureCSharpSymbolOrNull<ITypeSymbol, TypeSymbol>($"{nameof(memberTypes)}[{i}]");
+            }
+
+            var fields = memberTypes.Select((t, i) => new AnonymousTypeField(memberNames[i], Location.None, (TypeSymbol)t)).ToImmutableArray();
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs]: I think that `SelectAsArray` can be used here. Also, its function parameter can accept three parameters, which avoids creating a closure.

Here's an example `tupleTypes.SelectAsArray((t, i, v) => new BoundDeconstructValuePlaceholder(v[i].Syntax, t), variables)`

File: src/Compilers/VisualBasic/Portable/Compilation/VisualBasicCompilation.vb [hunk 3]
```diff
@@ -2605,6 +2605,24 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
...
+                memberNames As ImmutableArray(Of String)) As INamedTypeSymbol
+
+            Dim index = 0
+            For Each t In memberTypes
+                t.EnsureVbSymbolOrNothing(Of TypeSymbol)($"{NameOf(memberTypes)}({index})")
+
+                index = index + 1
+            Next
+
+            Dim fields = memberTypes.Select(Function(t, i) New AnonymousTypeField(memberNames(i), DirectCast(t, TypeSymbol), Location.None)).
```
Comment (jcouv) [inline src/Compilers/VisualBasic/Portable/Compilation/VisualBasicCompilation.vb]: Ditto here

File: src/Compilers/VisualBasic/Portable/Compilation/VisualBasicCompilation.vb [hunk 4]
```diff
@@ -2605,6 +2605,24 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
...
+                                     ToImmutableArray()
+
+            Dim descriptor = New AnonymousTypeDescriptor(fields, Location.None, isImplicitlyDeclared:=False)
```
Comment (jcouv) [inline src/Compilers/VisualBasic/Portable/Compilation/VisualBasicCompilation.vb]: I'm not familiar with `isImplicitlyDeclared`. How did you choose `false`?

File: src/Compilers/Core/Portable/Compilation/Compilation.cs [hunk 5]
```diff
@@ -851,6 +851,31 @@ public INamedTypeSymbol CreateTupleTypeSymbol(INamedTypeSymbol underlyingType, I
...
+                    throw new ArgumentNullException($"{nameof(memberTypes)}[{i}]");
+                }
+            }
+
+            return CommonCreateAnonymousTypeSymbol(memberTypes, memberNames);
+        }
+
+        protected abstract INamedTypeSymbol CommonCreateAnonymousTypeSymbol(
+            ImmutableArray<ITypeSymbol> memberTypes, ImmutableArray<string> memberNames);
+
```
Comment (jcouv) [inline src/Compilers/Core/Portable/Compilation/Compilation.cs]: Could you add compiler tests for the new APIs (VB and C#)?

For some inspiration, see `CreateTupleTypeSymbol_BadArguments` or `CreateTupleTypeSymbol_WithValueTuple`. 
I'm not sure which file they should go in though...

File: src/Compilers/Core/Portable/Compilation/Compilation.cs (hunk 1 above)
Comment (jcouv) [inline src/Compilers/Core/Portable/Compilation/Compilation.cs]: I'll let @gafter comment on that public API. 
I think it's fine the way you did it.

**Comment by jcouv** [comment]
Done with review pass. Main request is to add compiler-level tests.

File: src/Compilers/Core/Portable/Compilation/Compilation.cs [hunk 6]
```diff
@@ -851,6 +851,31 @@ public INamedTypeSymbol CreateTupleTypeSymbol(INamedTypeSymbol underlyingType, I
...
+                throw new ArgumentException($"{nameof(memberTypes)} and {nameof(memberNames)} must have the same length.");
+            }
+
+            for (int i = 0, n = memberTypes.Length; i < n; i++)
+            {
+                if (memberTypes[i] == null)
+                {
+                    throw new ArgumentNullException($"{nameof(memberTypes)}[{i}]");
+                }
+            }
```
Comment (cston) [inline src/Compilers/Core/Portable/Compilation/Compilation.cs]: Doesn't `CommonCreateAnonymousTypeSymbol` check this?

**Comment by CyrusNajmabadi** [comment]
@gafter can you take a look?  I'd like to make sure you're ok with this before investing in testing.

File: src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs]: awesome.  i'll make that change!

File: src/Compilers/Core/Portable/Compilation/Compilation.cs (hunk 5 above)
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Compilation/Compilation.cs]: Will do when i get signoff from one more person that hte API shape is ok!

File: src/Compilers/Core/Portable/Compilation/Compilation.cs (hunk 6 above)
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Compilation/Compilation.cs]: This loop checks for null.  THe individual language loops check that the ITypeSymbols are the right language Type-Symbols.  The function those loops call check if it's a "TypeSymbol *or* Null".  They don't throw in the 'null' case.  I could have each of those loops check for null.  Or i could just have the loop check here for both.

**Comment by CyrusNajmabadi** [comment]
Ok.  I added a couple of sanity tests for C# and VB.

File: src/Compilers/CSharp/Test/Symbol/Compilation/CompilationAPITests.cs [hunk 7]
```diff
@@ -2150,6 +2150,52 @@ public void PreviousSubmissionWithError()
...
+            var compilation = CSharpCompilation.Create("HelloWorld");
+            var type = compilation.CreateAnonymousTypeSymbol(
+                        ImmutableArray.Create<ITypeSymbol>(compilation.GetSpecialType(SpecialType.System_Int32), compilation.GetSpecialType(SpecialType.System_Boolean)),
+                        ImmutableArray.Create("m1", "m2"));
+
+            Assert.True(type.IsAnonymousType);
+            Assert.Equal(2, type.GetMembers().OfType<IPropertySymbol>().Count());
+            Assert.Equal("<anonymous type: int m1, bool m2>", type.ToDisplayString());
+        }
+
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Symbol/Compilation/CompilationAPITests.cs]: Thanks! 
I would also suggest a few more API tests: 

1. Pass 2 types, but only 1 name, or maybe 3 names, or some null names.
2. Pass some VB types into C#, or C# types into VB.

**Comment by CyrusNajmabadi** [comment]
More tests added.

**Comment by jcouv** [comment]
LGTM. Thanks for the extra tests!

File: src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs [hunk 8]
```diff
@@ -2835,6 +2835,20 @@ protected override INamedTypeSymbol CommonCreateTupleTypeSymbol(INamedTypeSymbol
...
+        {
+            for (int i = 0, n = memberTypes.Length; i < n; i++)
+            {
+                memberTypes[i].EnsureCSharpSymbolOrNull<ITypeSymbol, TypeSymbol>($"{nameof(memberTypes)}[{i}]");
+            }
+
+            var fields = memberTypes.SelectAsArray((type, index, loc) => new AnonymousTypeField(memberNames[index], loc, (TypeSymbol)type), Location.None);
+            var descriptor = new AnonymousTypeDescriptor(fields, Location.None);
+
+            return this.AnonymousTypeManager.ConstructAnonymousTypeSymbol(descriptor);
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs]: I'm concerned that calling this API may have the side-effect of changing the generated IL. I suspect it doesn't, but can we please verify that?

Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs]: sure.

File: src/Compilers/Core/Portable/Compilation/Compilation.cs (hunk 1 above)
Comment (gafter) [inline src/Compilers/Core/Portable/Compilation/Compilation.cs]: This API makes sense to me.  I think it makes sense for this and the tuple API to be done the same way.

File: src/Compilers/Core/Portable/Compilation/Compilation.cs [hunk 9]
```diff
@@ -851,6 +851,46 @@ public INamedTypeSymbol CreateTupleTypeSymbol(INamedTypeSymbol underlyingType, I
...
+
+            if (memberNames.IsDefault)
+            {
+                throw new ArgumentNullException(nameof(memberNames));
+            }
+
+            if (memberTypes.Length != memberNames.Length)
+            {
+                throw new ArgumentException($"{nameof(memberTypes)} and {nameof(memberNames)} must have the same length.");
```
Comment (gafter) [inline src/Compilers/Core/Portable/Compilation/Compilation.cs]: I think this would need to be internationalized.

**Comment by gafter** [comment]
The API shape LGTM

**Pull request merged**